src/Codegen/TypeDef.hs

```diff
@@ -1,10 +1,121 @@
 
+import           Common.Identifiers             ( DConId
+                                                , TConId
+                                                , fromString
+                                                , ident
+                                                )
+
+import           Codegen.Identifiers
+import           Data.Function                  ( on )
+import           Data.List as A                 ( maximumBy
+                                                , partition
+                                                )
+import Data.Map as M
+
+data TypeDefInfo = TypeDefInfo
+  { dconType :: M.Map String String
+  , typeSize :: M.Map String Int
+  , isPointer :: M.Map String Bool
+  , tag :: M.Map String Int
+  }
+
+instance Semigroup TypeDefInfo where
+  a <> b = combineTypeDefInfo a b
+
+instance Monoid TypeDefInfo where
+  mempty = TypeDefInfo empty empty empty empty
+
+combineTypeDefInfo :: TypeDefInfo -> TypeDefInfo -> TypeDefInfo
+combineTypeDefInfo a b = TypeDefInfo typ sz isPtr tags 
+ where
+   typ = dconType a `M.union` dconType b
+   sz = typeSize a `M.union` typeSize b
+   isPtr = isPointer a `M.union` isPointer b
+   tags = tag a `M.union` tag b
+
 -- | Generate definitions for SSM type definitions.
-genTypeDef :: L.TypeDef L.Type -> C.Definition
-genTypeDef = error "Not yet implemented"
+genTypeDef :: TConId -> L.TypeDef L.Type -> ([C.Definition], TypeDefInfo)
+genTypeDef tconid (L.TypeDef dCons _) = ([tags,structDef],info)
+ where
+  {- | Return the size in bits of a given field 
+  (assume all built-in types are one word for now)
+  -}
+  fieldSize :: L.Type -> Int
+  fieldSize (L.TBuiltin _) = 32
+  fieldSize (L.TCon     _) = 32
+
+  -- | Return the size in bits of a given data constructor
+  dConSize :: (DConId, L.TypeVariant L.Type) -> Int
+  dConSize (_, L.VariantNamed fields  ) = sum $ fieldSize . snd <$> fields
+  dConSize (_, L.VariantUnnamed fields) = sum $ fieldSize <$> fields
+
+  -- | Distinguish heap objects from possible int objects
+  (heapObjs, possibleIntObjs) = A.partition ((>= 32) . dConSize) dCons
+
+  -- | Incorporate tag, then partition list based on whether each 'DCon is still < 32 bits 
+  smallEnough
+    :: [(DConId, L.TypeVariant L.Type)]
+    -> ([(DConId, L.TypeVariant L.Type)], [(DConId, L.TypeVariant L.Type)])
+  smallEnough candidates = A.partition ((< 32) . dataBitsPlus tagBits) candidates
+   where
+    dataBitsPlus :: Int -> (DConId, L.TypeVariant L.Type) -> Int
+    dataBitsPlus bits dcon = bits + dConSize dcon
+    tagBits = q + r where (q, r) = length candidates `quotRem` 2
+
+  -- | Idenfity int objects by considering the space needed for tag bits
+  (moreHeapObs, intObjs) = smallEnough possibleIntObjs
+
+  -- | Arrange data constructors from smallest to largest tag value
+  tagged                 = intObjs ++ moreHeapObs ++ heapObjs
+
+  -- | Define enum name as 'TConId ++ Tag 
+  enumName               = CIdent $ fromString $ ident tconid ++ "Tag"
```

Comment:
No need to append `++ "Tag"` to the identifier; we'll handle name mangling elsewhere

Reply:
C doesn't let an enum and struct share the same identifier. If I have `data MyBool = MyTrue int | MyFalse int int `, once the C struct is named MyBool, I can't use MyBool again to name the enum. Should I ditch the ADT-specific C struct and only use generic ADT structs in initialization? This way the "MyBool" name would be available to use with the enum.